### ISSUE TITLE
JDK 9 via Homebrew Update breaks rJava

### ISSUE THREAD

**Issue by dantonnoriega** [s-u/rJava#119]
TL;DR

Installing JDK9 appears via homebrew appears to break `rJava`. Reverting back to JDK8 fixes the problem. Upon inspecting the not loaded dylib `libjvm.dylib`, the path to this lib appears to have changed between version (the `/jre/` does not exist in the JDK9 path).

```
~ ❯ ls /Library/Java/JavaVirtualMachines/jdk-9.jdk/Contents/Home/lib/server
Xusage.txt    libjsig.dylib libjvm.dylib
~ ❯ ls /Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home/jre/lib/server
Xusage.txt    libjsig.dylib libjvm.dylib
```

My hypothesis is that this may be the causing the problem for me after the upgrade? I know that seems too simple a reason, but keep in mind, I know nothing about Java.

Details of my adventure are below. My `sessionInfo()` and hardware profile can be found at the very, very bottom, in case that matters for some reason.

---

I use homebrew to install java. After an update today to JDK9, I received the following error when loading `rJava`:

```
> library(rJava)
Error: package or namespace load failed for ‘rJava’:
 .onLoad failed in loadNamespace() for 'rJava', details:
  call: dyn.load(file, DLLpath = DLLpath, ...)
  error: unable to load shared object '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/rJava/libs/rJava.so':
  dlopen(/Library/Frameworks/R.framework/Versions/3.4/Resources/library/rJava/libs/rJava.so, 6): Library not loaded: @rpath/libjvm.dylib
  Referenced from: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/rJava/libs/rJava.so
  Reason: image not found
```

Here is my JDK9 Install

```
~ ❯ /usr/libexec/java_home -V
Matching Java Virtual Machines (1):
    9, x86_64:	"Java SE 9"	/Library/Java/JavaVirtualMachines/jdk-9.jdk/Contents/Home

/Library/Java/JavaVirtualMachines/jdk-9.jdk/Contents/Home
```

Here is the results of running `R CMD javareconf`. Compared to past observations, `$(JAVA_HOME)` does not make a strong appearance. And `Java library path:` remains blank.
```
~ ❯ R CMD javareconf
Java interpreter : /usr/bin/java
Java version     : 9
Java home path   : /Library/Java/JavaVirtualMachines/jdk-9.jdk/Contents/Home
Java compiler    : /usr/bin/javac
Java headers gen.: /usr/bin/javah
Java archive tool: /usr/bin/jar
System Java on macOS

trying to compile and link a JNI program 
detected JNI cpp flags    : -I/System/Library/Frameworks/JavaVM.framework/Headers
detected JNI linker flags : -framework JavaVM
gcc-7 -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG -I/System/Library/Frameworks/JavaVM.framework/Headers  -I/usr/local/include   -fPIC  -Wall -g -O2  -c conftest.c -o conftest.o
conftest.c: In function 'main':
conftest.c:4:5: warning: 'JNI_CreateJavaVM' is deprecated [-Wdeprecated-declarations]
     JNI_CreateJavaVM(0, 0, 0);
     ^~~~~~~~~~~~~~~~
In file included from conftest.c:1:0:
/System/Library/Frameworks/JavaVM.framework/Headers/jni.h:1937:1: note: declared here
 JNI_CreateJavaVM(JavaVM **pvm, void **penv, void *args);
 ^~~~~~~~~~~~~~~~
gcc-7 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/Library/Frameworks/R.framework/Resources/lib -L/usr/local/lib -o conftest.so conftest.o -framework JavaVM -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation


JAVA_HOME        : /Library/Java/JavaVirtualMachines/jdk-9.jdk/Contents/Home
Java library path: 
JNI cpp flags    : -I/System/Library/Frameworks/JavaVM.framework/Headers
JNI linker flags : -framework JavaVM
Updating Java configuration in /Library/Frameworks/R.framework/Resources
Done.
```

I then attempted the "solution" from this [SO post](https://stackoverflow.com/questions/30738974/rjava-load-error-in-rstudio-r-after-upgrading-to-osx-yosemite):

```
sudo ln -f -s $(/usr/libexec/java_home)/jre/lib/server/libjvm.dylib /usr/local/lib
```

But received an error. *The path apparently does not exist for this homebrew install of JDK9*.

```
~ ❯ ls $(/usr/libexec/java_home)/jre/lib/server/

ls: /Library/Java/JavaVirtualMachines/jdk-9.jdk/Contents/Home/jre/lib/server/: No such file or directory
```

I noticed that the `jre` directory no longer exist, instead finding the `libjvm.dylib` located here

```
~ ❯ ls /Library/Java/JavaVirtualMachines/jdk-9.jdk/Contents/Home/lib/server
Xusage.txt    libjsig.dylib libjvm.dylib
```

I attempted the symlink solution again, but fixed the path. This did not solve the problem.

In the end, I removed JDK9, and reinstalled JDK8 via homebrew and reconfigured. This fixed the problem and rJava worked good as new.

```
~ ❯ brew cask install java8
~ ❯ R CMD javareconf
Java interpreter : /usr/bin/java
Java version     : 1.8.0_144
Java home path   : /Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home/jre
Java compiler    : /usr/bin/javac
Java headers gen.: /usr/bin/javah
Java archive tool: /usr/bin/jar
Non-system Java on macOS

trying to compile and link a JNI program 
detected JNI cpp flags    : -I$(JAVA_HOME)/../include -I$(JAVA_HOME)/../include/darwin
detected JNI linker flags : -L$(JAVA_HOME)/lib/server -ljvm
gcc-7 -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG -I/Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home/jre/../include -I/Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home/jre/../include/darwin  -I/usr/local/include   -fPIC  -Wall -g -O2  -c conftest.c -o conftest.o
gcc-7 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/Library/Frameworks/R.framework/Resources/lib -L/usr/local/lib -o conftest.so conftest.o -L/Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home/jre/lib/server -ljvm -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation


JAVA_HOME        : /Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home/jre
Java library path: $(JAVA_HOME)/lib/server
JNI cpp flags    : -I$(JAVA_HOME)/../include -I$(JAVA_HOME)/../include/darwin
JNI linker flags : -L$(JAVA_HOME)/lib/server -ljvm
Updating Java configuration in /Library/Frameworks/R.framework/Resources
Done.
```

---

Possibly relevant R and system info.

```
> sessionInfo()
R version 3.4.1 (2017-06-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] magrittr_1.5   colorout_1.1-2

loaded via a namespace (and not attached):
[1] compiler_3.4.1
```

```
~ ❯ system_profiler SPHardwareDataType
    Hardware Overview:

      Model Name: MacBook Pro
      Model Identifier: MacBookPro11,1
      Processor Name: Intel Core i5
      Processor Speed: 2.4 GHz
      Number of Processors: 1
      Total Number of Cores: 2
      L2 Cache (per Core): 256 KB
      L3 Cache: 3 MB
      Memory: 8 GB
```

**Comment by luisspuerto** [comment]
I think I have the same problem... any solution?

**Comment by luisspuerto** [comment]
I can confirm that uninstalling Java 9 and reinstalling and configuring Java 8 works. Is there anyway where we can have both installations tell R and rJava, or at least rJava to use Java 8 while the rest of the system use Java 9? 

Thanks!

**Comment by luisspuerto** [comment]
Besides, a new R version has been just released. Let's pray that nothing more will break. 🤕

**Comment by luisspuerto** [comment]
With the rJava update this issue at least for me is fixed

**Comment by s-u** [comment]
Thanks!

**Comment by AxDef** [comment]
hello, 

I have the same problem with jdk9 and rJava, it is not working. What did you do to fix it ? What do you mean by the : "the rJava update" ? I installed R3.4.2 and it doesn't work.

**Comment by s-u** [comment]
You must re-configure R after Java updates, because Oracle changes the paths with every upgrade (utterly stupid, but that's Oracle - unfortunately Apple dropped the ball).

**Comment by luisspuerto** [comment]
@AxDef rJava was update a little bit latter of I had the issue and with the update the issue was fixed. If you have problems, I recommend you to reconfigure R with Java as @s-u recommended.
```
R CMD javareconf`
```
or 
```
R CMD javareconf JAVA_CPPFLAGS=-I/System/Library/Frameworks/JavaVM.framework/Headers`
```
usually works.

**Comment by AxDef** [comment]
Many thanks for your help, yet it is still not working : 

conftest.c:4:5: warning: 'JNI_CreateJavaVM' is deprecated [-Wdeprecated-declarations]
    JNI_CreateJavaVM(0, 0, 0);

**Comment by AxDef** [comment]
[java reconf.txt](https://github.com/s-u/rJava/files/1520866/java.reconf.txt)
[rJava .:config.txt](https://github.com/s-u/rJava/files/1520867/rJava.config.txt)

**Comment by luisspuerto** [comment]
When are you seeing that message? 

do you have a homebrew install and you compile R and rJava or do you have a CRAN install?

**Comment by AxDef** [comment]
[java reconf.txt](https://github.com/s-u/rJava/files/1520875/java.reconf.txt)

**Comment by AxDef** [comment]
[rJava .:config.txt](https://github.com/s-u/rJava/files/1520876/rJava.config.txt)

**Comment by AxDef** [comment]
[rJava .:config.txt](https://github.com/s-u/rJava/files/1520881/rJava.config.txt)

**Comment by luisspuerto** [comment]
I think I have the same output and mine work without problem. 

```
sudo R CMD javareconf \                                                                  10:32:16  01.12.17
    JAVA_CPPFLAGS=-I/System/Library/Frameworks/JavaVM.framework/Headers

Password:
Java interpreter : /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/bin/java
Java version     : 9.0.1
Java home path   : /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
Java compiler    : /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/bin/javac
Java headers gen.: /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/bin/javah
Java archive tool: /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/bin/jar
Non-system Java on macOS

trying to compile and link a JNI program
detected JNI cpp flags    : -I/System/Library/Frameworks/JavaVM.framework/Headers
detected JNI linker flags : -L$(JAVA_HOME)/lib/server -ljvm
/usr/local/opt/llvm/bin/clang -I/usr/local/Cellar/r/3.4.3/lib/R/include -DNDEBUG -I/System/Library/Frameworks/JavaVM.framework/Headers  -I/usr/local/opt/gettext/include -I/usr/local/opt/llvm/include   -fPIC  -g -O3 -Wall -pedantic -std=gnu99 -mtune=native -pipe -c conftest.c -o conftest.o
conftest.c:4:5: warning: 'JNI_CreateJavaVM' is deprecated [-Wdeprecated-declarations]
    JNI_CreateJavaVM(0, 0, 0);
    ^
/System/Library/Frameworks/JavaVM.framework/Headers/jni.h:1936:39: note: 'JNI_CreateJavaVM' has been explicitly marked
      deprecated here
_JNI_IMPORT_OR_EXPORT_ __attribute__((deprecated)) jint JNICALL
                                      ^
1 warning generated.
/usr/local/opt/llvm/bin/clang -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/usr/local/Cellar/r/3.4.3/lib/R/lib -L/usr/local/opt/gettext/lib -L/usr/local/opt/llvm/lib -Wl,-rpath,/usr/local/opt/llvm/lib -o conftest.so conftest.o -L/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/lib/server -ljvm -L/usr/local/Cellar/r/3.4.3/lib/R/lib -lR -lintl -Wl,-framework -Wl,CoreFoundation


JAVA_HOME        : /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
Java library path: $(JAVA_HOME)/lib/server
JNI cpp flags    : -I/System/Library/Frameworks/JavaVM.framework/Headers
JNI linker flags : -L$(JAVA_HOME)/lib/server -ljvm
Updating Java configuration in /usr/local/Cellar/r/3.4.3/lib/R
Done.
```

**Comment by luisspuerto** [comment]
I just installed everything from scratch and in a clean install yesterday in a "old" mac at home and right now I just updated to R 3.4.3. the I've done a install with homebrew with openblas and openmp support and so, and I have no problems at all. I haven't tested everything, but everything load just fine. 

Right now I'm updating to R 3.4.3 on my computer...

**Comment by AxDef** [comment]
thanks

I'll try R3.4.3


Now when I open R, install.packages("rJava") from source is not working.

**Comment by AxDef** [comment]
My initial issue was to fix this : 

dyld: Library not loaded: @rpath/libjvm.dylib

**Comment by gwinstanley** [comment]
I wrote a small [helper page](http://www.snaq.net/software/rjava-macos.php) about this recently, which may be useful for anyone landing here.

**Comment by AxDef** [comment]
Thanks !

**Comment by luisspuerto** [comment]
by the way if I configure: 

```
sudo R CMD javareconf \ 
JAVA_CPPFLAGS='-I/System/Library/Frameworks/JavaVM.framework/Headers -I/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/'
```

I get no warning: 
```
/usr/local/Cellar/r/3.4.3/lib/R/bin/javareconf: line 66: -I/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/: No such file or directory
Java interpreter : /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/bin/java
Java version     : 9.0.1
Java home path   : /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
Java compiler    : /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/bin/javac
Java headers gen.: /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/bin/javah
Java archive tool: /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/bin/jar
Non-system Java on macOS

trying to compile and link a JNI program
detected JNI cpp flags    : -I$(JAVA_HOME)/include -I$(JAVA_HOME)/include/darwin
detected JNI linker flags : -L$(JAVA_HOME)/lib/server -ljvm
/usr/local/opt/llvm/bin/clang -I/usr/local/Cellar/r/3.4.3/lib/R/include -DNDEBUG -I/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/include -I/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/include/darwin  -I/usr/local/opt/gettext/include -I/usr/local/opt/llvm/include   -fPIC  -g -O3 -Wall -pedantic -std=gnu99 -mtune=native -pipe -c conftest.c -o conftest.o
/usr/local/opt/llvm/bin/clang -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/usr/local/Cellar/r/3.4.3/lib/R/lib -L/usr/local/opt/gettext/lib -L/usr/local/opt/llvm/lib -Wl,-rpath,/usr/local/opt/llvm/lib -o conftest.so conftest.o -L/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/lib/server -ljvm -L/usr/local/Cellar/r/3.4.3/lib/R/lib -lR -lintl -Wl,-framework -Wl,CoreFoundation


JAVA_HOME        : /Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home
Java library path: $(JAVA_HOME)/lib/server
JNI cpp flags    : -I$(JAVA_HOME)/include -I$(JAVA_HOME)/include/darwin
JNI linker flags : -L$(JAVA_HOME)/lib/server -ljvm
Updating Java configuration in /usr/local/Cellar/r/3.4.3/lib/R
Done.
```

**Comment by luisspuerto** [comment]
I also have set in `.bash_profile` & `.zshrc` $JAVA_HOME as

`export JAVA_HOME="$(/usr/libexec/java_home)"`

**Comment by AxDef** [comment]
Sorry I followed the instructions in the "helper page" above, yet I still have : 

ERROR: compilation failed for package ‘rJava’
* removing ‘/Library/Frameworks/R.framework/Versions/3.4/Resources/library/rJava’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/3.4/Resources/library/rJava’

**Comment by AxDef** [comment]
[fix-rJava.sh.txt](https://github.com/s-u/rJava/files/1521311/fix-rJava.sh.txt)

**Comment by gwinstanley** [comment]
I notice in the R v3.4.3 changes they've made some amendments to the ```javareconf``` script, which may be causing additional problems. My page was written using v3.4.2 on macOS Sierra, so seems like things have changed, yet again. I'll see if I can update the script to work again for v3.4.3

**Comment by AxDef** [comment]
Thank you, I also tried to uninstall all and go back to previous versions of R and JDK like in : 

https://github.com/snowflakedb/dplyr-snowflakedb/wiki/Configuring-R-rJava-RJDBC-on-Mac-OS-X

and

http://www.owsiak.org/r-java-rjava-and-macos-adventures/

But still not working. 


Thanks.

**Comment by gwinstanley** [comment]
Ok, seems they've finally got it right in R 3.4.3 and recognise that Java >6 is not system-installed. So, it's not much simpler, and is just the following:

1. Make sure you have JDK 9.0.1 installed.
2. If you're installing via Homebrew, make sure your Terminal session has JAVA_HOME set correctly for your JDK 9.0.1 installation.
3. Install R (either from regular installer package, or with Homebrew and the "--with-java" option).
4. (Re)install rJava from within R.

That just worked fine for me on a Homebrew installation on macOS 10.13.1 (High Sierra).

**Comment by AxDef** [comment]
Many thanks, finally, it works. I did the installation of rJava with R studio.

Regards
